### ISSUE TITLE
RHS of 'noremap' mappings are passed to popup filters

### ISSUE THREAD

**Issue by bfrg** [vim/vim#7459]
**Describe the bug**
The RHS of `noremap` mappings are forwarded to popup filters when the popup entry `mapping` is enabled.

I don't know if that's intentional or a bug. But for me this is unexpected because `noremap` mappings are non-recursive. Passing the RHS to a popup filter can result in very awkward behavior.

**To Reproduce**
1. Run `vim --clean -S test.vim`, with
   ```vim
   nnoremap <c-n> <esc>itest

   function PopupFilter(winid, key) abort
       if a:key ==# "\<esc>"
           call popup_close(a:winid)
           return v:true
       endif
       return v:false
   endfunction

   call popup_create('test test test test', {
           \ 'border': [],
           \ 'filter': 'PopupFilter',
           \ 'filtermode': 'n',
           \ 'mapping': v:true
           \ })
   ```
2. Press `CTRL-N`.
3. Result: The popup window is closed and "test" is inserted into the buffer.

**Expected behavior**
The RHS of a mapping - in particular of a `noremap` - should not be passed to a popup-filter.

I expect `noremap` mappings to always work the same, independent of whether a popup window with a popup-filter is displayed.

**Environment:**
Vim version `8.2.2127`

**Comment by lacygoill** [comment]
I agree that the current design sometimes leads to unexpected behaviors.

I don't know how mappings and popup filters are implemented, but I *think* the former are processed before the latter.  That would explain what you [and I](https://github.com/lacygoill/wiki/blob/master/vim/popup.md#in-which-order-are-mappings-and-filters-processed) observe.

I don't know if it can easily be fixed, but as a temporary workaround, we can install a temporary buffer-local mapping which replaces the lhs with itself.  For example:
```vim
nnoremap <c-n> <esc>itest
nnoremap <buffer><nowait> <c-n> <c-n>

function PopupFilter(winid, key) abort
    if a:key ==# "\<esc>"
        call popup_close(a:winid)
        return v:true
    endif
    return v:false
endfunction

function s:Callback(...) abort
    nunmap <buffer> <c-n>
endfunction

call popup_create('test test test test', {
        \ 'border': [],
        \ 'filter': 'PopupFilter',
        \ 'filtermode': 'n',
        \ 'mapping': v:true,
        \ 'callback': function('s:Callback'),
        \ })
```
I documented how it works [here](https://github.com/lacygoill/wiki/blob/master/vim/popup.md#my-filter-doesnt-work--when-i-press-a-key-its-not-handled-as-expected), and I use it [here](https://github.com/lacygoill/vim-qf/blob/2fe356db5d7cf3c8cce1ad24d619873a2fad13ac/autoload/qf/preview.vim#L94).

But it's far from perfect.  It doesn't scale well for a large number of mappings, and might destroy existing buffer-local mappings (so you might want to save/restore buffer-local mappings instead of just running an `:unmap`).

In the previous example, there exists a simpler workaround, which is to avoid `<esc>` in the rhs, and use `<c-\><c-n>`:

    nnoremap <c-n> <esc>itest
                   ^---^
                     ✘

    nnoremap <c-n> <c-\><c-n>itest
                   ^--------^
                       ✔

Which is better [because of this](https://github.com/lacygoill/wiki/blob/master/vim/mapping.md#what-if-esc-is-used-in-the-rhs-of-a-mapping).

Anyway, it would be nice if it could be fixed in Vim and we didn't have to work around the issue.

**Comment by lacygoill** [comment]
> It doesn't scale well for a large number of mappings, and might destroy existing buffer-local mappings (so you might want to save/restore buffer-local mappings instead of just running an :unmap).

And it doesn't work if the key results from the expansion of a mapping (rather than from an interactive keypress).

**Comment by bfrg** [comment]
Thanks for the thorough explanation.

From the link above:
> In which order are mappings and filters processed?
>
>The mappings are processed before the filters.

There's the possibility to disable mappings by setting the `mapping` entry to `false`.
```vim
call popup_create('test test test test', {
        \ 'border': [],
        \ 'filter': 'PopupFilter',
        \ 'filtermode': 'n',
        \ 'mapping': v:false
        \ })
```
However, now none of the user-defined mappings will work as long as the popup window is displayed. Unfortunately, there's no in-between. See  #4759.

**Comment by brammool** [comment]
I don't see a direct relation between keys that come from the user / mapping / noremap  and the filter.  The filter gets all keys, no matter whether they were mapped or not.
It can even be useful: if a popup from a plugin requires F7, the user can map another key to F7 to his liking.
If you want to check for Esc being pressed, you could perhaps use getchar(1) if there is any input following.

**Comment by bfrg** [comment]
>  It can even be useful: if a popup from a plugin requires F7, the user can map another key to F7 to his liking.

Okay, I think I misunderstood the design of popup filters. That's certainly a nice way to remap a popup key. But it also means that users can never have F7 on the RHS in _any_ of their mappings, unless it's to invoke a filter function. If, however, someone has F7 somewhere on the RHS of a mapping, that is completely unrelated to the popup, it will break that mapping.

I always assumed plugins would provide a way to override the default popup key through a variable (at least that's what I did so far). Something like:
```vim
function PopupFilter(winid, key) abort
    if a:key ==# get(g:, 'myplugin_popup_close', "\<esc>")
        call popup_close(a:winid)
        return v:true
    endif
    return v:false
endfunction
```

The way it is implemented now means that the RHS of a `noremap` should never contain any keys that a popup filter might swallow, otherwise it will break the mapping.

Maybe I'm just over-complicating things here. I'm curious to see what other users have to say about this and have done so far.
